README
GNU GPL v2

Ce logiciel baisse le volume (à 5%) de votre ordinateur durant les pubs vocales spotify,
en détectant le titre de la chanson dans la fenêtre wine.
Un contre système remonte légèrement le son au besoin, pour éviter la mise en pause automatique de la pub.


CHANGELOG
04/10/2009:	Ne bloque plus le niveau du volume. Le volume ne changement plus quand le seuil est atteint. 


02/10/2009: 	Conserve le niveau du son choisi par l'utilisateur (donc ne remonte plus à 100% après une pub)
		Sortie de la boucle quand spotify est fermé 
		Nouveau mixer + standard - test des logiciels requis
		

30/09/2009: 	fonctionne maintenant sans seuil empirique et ajout d'un tag de déclenchement (tagueule). 50% plus rapide (moins de pause automatique de spotify)


29/09/2009: 	plus de crossfader (pub coupée cash), plus le bug de son faible à demarrage.


28/09/2009: 	la liste des pubs est auto-updatée via une url


27/09/2009: 	fonctionne sur une liste pubs, plus simplement sur un seul pattern 


spotify.sh
#!/bin/bash
###
### Under GPL v2 - pronoiaque@gmail.com
### thx aidos
###
### Variables et lancement de Spotify
###
declare -i volume
declare -i tagueule

tagueule=0
user=`whoami`

###################################
###  On récupère la liste des Pubs
###
### ( Merci de me signaler s'il manque des pubs, si possible avec leur titre )
###
if [ $1 ] ; then
	pubpatterns=$(wget -O - "`echo $1`" 2>/dev/null)
  else
	pubpatterns=$(wget -O - http://github.com/pronoiaque/Spotify.sh/raw/master/spotify.pub 2>/dev/null)

fi

##################################################################"
### Est-ce que les packages necessaires sont installés ?

if [ ! -x /usr/bin/wmctrl ] ; then
        zenity --question --text "Le package <b>wmctrl</b> n'est pas detecté.\nVoulez-vous l'installer ?"
        if [ $? == "0" ] ; then
                gksudo aptitude install wmctrl
        fi
fi

if [ ! -x /usr/bin/amixer ] ; then
        zenity --question --text "Le package <b>alsa-utils</b> n'est pas detecté.\nVoulez-vous l'installer ?"
        if [ $? == "0" ] ; then
                gksudo aptitude install alsa-utils
        fi
fi

####################################
### On lance Spotify, avant toute chose
###
env WINEPREFIX="/home/$user/.wine" wine 2>/dev/null "C:\Program Files\Spotify\spotify.exe"&


########################################
### On crée un fonction qui détecte le titre de la fenêtre Spotify
### et qui crée un "ads" non NULL si un pub est détectée

function grab_titre
{
	titre=$(echo `wmctrl -l | grep Spotify | cut -d" " -f 5-500`)
	ads=$(echo `echo $titre | grep -iE "$pubpatterns"`)
}

######################################
###  Une pour recuperer les valeurs du volume (Master + PCM ) de l'utilisateur
###
function get_user_vol
{
	VolMaster=`amixer -c 0 cget name='Master Playback Volume' | grep : | sed 's/^.*=\([^,]*\).*$/\1/'`
	VolPCM=`amixer -c 0 cget name='PCM Playback Volume' | grep : | sed 's/^.*=\([^,]*\).*$/\1/'`
}

################################################
### Pour les reinjecter après une pub
function put_user_vol
{
	amixer -q cset name='Master Playback Volume' $VolMaster
	amixer -q cset name='PCM Playback Volume' $VolPCM
}

#############################################
### Pour Regler le Mixer PCM à une certain valeur
###
function setvolume
{
	amixer -c 0 cset name='PCM Playback Volume' $1 1>/dev/null
}

###########################################
### le Volume PCM sert de 1 niveau
###
get_user_vol
volume=$VolPCM

###################################################
### Début de la Boucle - temps que Spotify tourne
while [ ! "`ps x | grep spotify.exe | grep -v grep`"  = "" ] ;
  do

	###############################
	## Recuperation titre et volume
	##
	grab_titre

	##############################################################################
	### Si une pub en PLAY est detectée avec un volume fort
	### - On baisse le volume au mininum -> Mise en PAUSE
	### - Ajout d'un trigger (ferme) "tagueule"

	if [ ! "$ads" = "" ] && [ ! "$titre" = "Spotify" ] && [ $volume -gt 7 ] ; then
		volume=1
		setvolume $volume
		sleep 0.4
		tagueule=1
		grab_titre
	fi

	#####################################################
	### Si Spotify est en PAUSE + trigger
	### - On augmente le volume jusqu`à la mise en PLAY
	###
	while [ "$titre" = "Spotify" ] && [ $tagueule = 1 ] ;
	 do
		volume=$((`echo $volume` + 3))
		setvolume $volume
		sleep 0.5
		grab_titre
	done

	##################################################################################################
	## Si la pub est en PLAY à un volume reduit
	## - On la laisse tourner
	##
	while [ ! "$ads" = "" ] && [ ! "$titre" = "Spotify" ] && [ $tagueule = 1 ] && [ $volume -le 7 ] ;
	 do
		grab_titre
		sleep 0.5
	done

	###############################################
	## S'il n'y a plus de pub
	## - On remets le

	if [ "$ads" = "" ] && [ $tagueule = 1 ] ; then
		put_user_vol
		tagueule=0
	fi

	################################
	## Recuperation du Volume
	##
	if [ $tagueule = 0 ] ; then
		get_user_vol
		volume=$VolPCM
	fi

	sleep 0.2
done
exit 0


spotify_old.sh
#!/bin/bash
###
### Under GPL v2 - pronoiaque@gmail.com
### thx aidos
###
### Variables et lancement de Spotify
###
declare -i volume
declare -i tagueule
declare -i adsurl

pubpatterns=$(wget -O - http://github.com/pronoiaque/Spotify.sh/raw/master/spotify.pub 2>/dev/null)
tagueule=""
user=`whoami`

env WINEPREFIX="/home/$user/.wine" wine 2>/dev/null "C:\Program Files\Spotify\spotify.exe"&



###################################
###  On récupère la liste des Pubs
###
### ( Merci de me signaler s'il manque des pubs, si possible avec leur titre )
###
if [ $1 ] ; then
	echo $1
	pubpatterns=$(wget -O - "`echo $1`" 2>/dev/null)
fi

##################################################################"
### Est-ce que les packages necessaires sont installés ?

if [ ! -x /usr/bin/wmctrl ] ; then
        zenity --question --text "Le package <b>wmctrl</b> n'est pas detecté.\nVoulez-vous l'installer ?"
        if [ $? == "0" ] ; then
                gksudo aptitude install wmctrl
        fi
fi

if [ ! -x /usr/bin/amixer ] ; then
        zenity --question --text "Le package <b>alsa-utils</b> n'est pas detecté.\nVoulez-vous l'installer ?"
        if [ $? == "0" ] ; then
                gksudo aptitude install alsa-utils
        fi
fi

####################################
### On lance Spotify, avant toute chose
###

`$WINEPREFIX`




########################################
### On crée un fonction qui détecte le titre de la fenêtre Spotify
### et qui crée un "tilt" si un pub est détectée

function grab_titre
{
	titre=$(echo `wmctrl -l | grep Spotify | cut -d" " -f 5-500`)
	tilt=$(echo `echo $titre | grep -iE "$pubpatterns"`)
}

######################################
###  Une pour recuperer les valeurs du volume (Master + PCM ) de l'utilisateur
###
function get_user_vol
{
	VolMaster=`amixer -c 0 cget name='Master Playback Volume' | grep : | sed 's/^.*=\([^,]*\).*$/\1/'`
	VolPCM=`amixer -c 0 cget name='PCM Playback Volume' | grep : | sed 's/^.*=\([^,]*\).*$/\1/'`
}
################################################
### Pour les reinjecter après une pub
function put_user_vol
{
	amixer -q cset name='Master Playback Volume' $VolMaster
	amixer -q cset name='PCM Playback Volume' $VolPCM
}

#############################################
### Pour Regler le Mixer PCM à une certain valeur
###
function setvolume
{
	amixer -c 0 cset name='PCM Playback Volume' $1
}

###########################################
### le Volume PCM sert de 1 niveau
###
get_user_vol
volume=$VolPCM

###################################################
### Début de la Boucle - temps que Spotify tourne
while [ ! "`ps x | grep spotify.exe | grep -v grep`"  = "" ] ;
  do

    grab_titre

       #########################################
       #### Pour le cas de la première pub détectée
       ####
    if [ ! "$tilt" = "" ] && [ ! "$titre" = "Spotify" ] && [ $volume -gt 5 ] ; then
        volume=1
        setvolume $volume 2>/dev/null
        sleep 0.5
        tagueule=1
        grab_titre
    fi


        ##########################################################
        #### Si une pub est détectée et qu'elle est mise en pause à cause de la baisse du volume
        #### - On remonte de deux crans et si elle n'est plus en pause, on la laisse se diffuser (sleep 5)

    while [ "$titre" = "Spotify" ] && [ $volume -lt $VolPCM ] && [ $tagueule = 1 ] ;
     do
        volume=volume+2
        setvolume $volume 2>/dev/null
        grab_titre
        if [ ! "$titre" = "Spotify" ] ; then
             sleep 5
          else
            sleep 0.2
        fi
    done

        ###########################################################
        ###
        ###

    while [ ! "$tilt" = "" ] && [ ! "$titre" = "Spotify" ] && [ $tagueule = 1 ] && [ $volume -gt 5 ] ;
     do
        volume=1
        setvolume $volume 2>/dev/null
        sleep 0.2
        grab_titre
    done



    ##############################################################
    ### Si il n'y a pas de pub, pas de pause et le son même légèrement baissé
    ### - on remonte le son au niveau du volume de l'utilisateur
    ###

    if  [ "$tilt" = "" ] && [ ! "$titre" = "Spotify" ] && [ $volume -lt $VolPCM ] ; then
        put_user_vol
        tagueule=""
	echo  "put user vol"
    fi

    ################################################################
    ## Pour pas bouffer toutes les ressources avec la boucle
    ## - On en profite pour check le volume choisi par l'utilisateur
    ##
    if [ "$tilt" = "" ] && [ ! "$titre" = "Spotify" ] && [ $volume -ge $VolPCM ] ; then
        sleep 0.2
        get_user_vol
   fi

  sleep 0.2

done
exit 0


spotify_beta.sh
#!/bin/bash
###
### Under GPL v2 - pronoiaque@gmail.com
### thx aidos
###
### Variables et lancement de Spotify
###
declare -i volume
declare -i tagueule

tagueule=0
user=`whoami`

###################################
###  On récupère la liste des Pubs
###
### ( Merci de me signaler s'il manque des pubs, si possible avec leur titre )
###
if [ $1 ] ; then
	pubpatterns=$(wget -O - "`echo $1`" 2>/dev/null)
  else
	pubpatterns=$(wget -O - http://github.com/pronoiaque/Spotify.sh/raw/master/spotify.pub 2>/dev/null)

fi

##################################################################"
### Est-ce que les packages necessaires sont installés ?

if [ ! -x /usr/bin/wmctrl ] ; then
        zenity --question --text "Le package <b>wmctrl</b> n'est pas detecté.\nVoulez-vous l'installer ?"
        if [ $? == "0" ] ; then
                gksudo aptitude install wmctrl
        fi
fi

if [ ! -x /usr/bin/amixer ] ; then
        zenity --question --text "Le package <b>alsa-utils</b> n'est pas detecté.\nVoulez-vous l'installer ?"
        if [ $? == "0" ] ; then
                gksudo aptitude install alsa-utils
        fi
fi

####################################
### On lance Spotify, avant toute chose
###
env WINEPREFIX="/home/$user/.wine" wine 2>/dev/null "C:\Program Files\Spotify\spotify.exe"&


########################################
### On crée un fonction qui détecte le titre de la fenêtre Spotify
### et qui crée un "ads" non NULL si un pub est détectée

function grab_titre
{
	titre=$(echo `wmctrl -l | grep Spotify | cut -d" " -f 5-500`)
	ads=$(echo `echo $titre | grep -iE "$pubpatterns"`)
}

######################################
###  Une pour recuperer les valeurs du volume (Master + PCM ) de l'utilisateur
###
function get_user_vol
{
	VolMaster=`amixer -c 0 cget name='Master Playback Volume' | grep : | sed 's/^.*=\([^,]*\).*$/\1/'`
	VolPCM=`amixer -c 0 cget name='PCM Playback Volume' | grep : | sed 's/^.*=\([^,]*\).*$/\1/'`
}

################################################
### Pour les reinjecter après une pub
function put_user_vol
{
	amixer -q cset name='Master Playback Volume' $VolMaster
	amixer -q cset name='PCM Playback Volume' $VolPCM
}

#############################################
### Pour Regler le Mixer PCM à une certain valeur
###
function setvolume
{
	amixer -c 0 cset name='PCM Playback Volume' $1 1>/dev/null
}

###########################################
### le Volume PCM sert de 1 niveau
###
get_user_vol
volume=$VolPCM

###################################################
### Début de la Boucle - temps que Spotify tourne
while [ ! "`ps x | grep spotify.exe | grep -v grep`"  = "" ] ;
  do

	###############################
	## Recuperation titre et volume
	##
	grab_titre

	##############################################################################
	### Si une pub en PLAY est detectée avec un volume fort
	### - On baisse le volume au mininum -> Mise en PAUSE
	### - Ajout d'un trigger (ferme) "tagueule" 

	if [ ! "$ads" = "" ] && [ ! "$titre" = "Spotify" ] && [ $volume -gt 6 ] ; then
		volume=1
		setvolume $volume
		sleep 0.2
		tagueule=1
		grab_titre
		echo "mise en PAUSE"
	fi

	#####################################################
	### Si Spotify est en PAUSE + trigger
	### - On augmente le volume jusqu`à la mise en PLAY
	###
	while [ "$titre" = "Spotify" ] && [ $tagueule = 1 ] ;
	 do
		volume=$((`echo $volume` + 3))
		setvolume $volume
		sleep 0.2
		grab_titre
		echo "Augmentation +3 du volume, jusquà mise en PLAY" "("$volume")"
	done

	##################################################################################################
	## Si la pub est en PLAY à un volume reduit
	## - On la laisse tourner
	##
	while [ ! "$ads" = "" ] && [ ! "$titre" = "Spotify" ] && [ $tagueule = 1 ] && [ $volume -le 6 ] ;
	 do
		grab_titre
		echo dodo
		sleep 0.5
	done

	###############################################
	## S'il n'y a plus de pub
	## - On remets le 

	if [ "$ads" = "" ] && [ $tagueule = 1 ] ; then
		put_user_vol
		tagueule=0
		echo "fin de la pub"
	fi

	################################
	## Recuperation du Volume
	##
	if [ $tagueule = 0 ] ; then
		get_user_vol
	fi

	sleep 0.2

done
exit 0
